src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +383,299 @@ public static void EnsureSocketsAreInitialized()
+                }
+
+                AddressInfoEx* result = context->Result;
+                string canonicalName = null;
+
+                List<IPAddress> addresses = new List<IPAddress>();
+
+                while (result != null)
+                {
+                    if (canonicalName == null && result->ai_canonname != IntPtr.Zero)
+                        canonicalName = Marshal.PtrToStringUni(result->ai_canonname);
+
+                    IPAddress ipAddress = null;
+
+                    if (result->ai_family == AddressFamily.InterNetwork)
+                    {
+                        ipAddress = CreateIPv4Address(result->ai_addr, result->ai_addrlen);
+                    }
+                    else if (SocketProtocolSupportPal.OSSupportsIPv6 && result->ai_family == AddressFamily.InterNetworkV6)
+                    {
+                        ipAddress = CreateIPv6Address(result->ai_addr, result->ai_addrlen);
+                    }
+
+                    if (ipAddress != null)
+                        addresses.Add(ipAddress);
+
+                    result = result->ai_next;
+                }
+
+                if (canonicalName == null)
+                    canonicalName = state.HostAddress;
+
+                state.CompleteAsyncResult(new IPHostEntry
+                {
+                    HostName = canonicalName,
+                    Aliases = Array.Empty<string>(),
+                    AddressList = addresses.ToArray()
+                });
+            }
+            finally
+            {
+                state.Dispose();
+            }
+        }
+
+        private static unsafe IPAddress CreateIPv4Address(byte* socketAddress, int addressLength)
+        {
+            const int IPv4SocketAddressSize = 16;
+
+            if (addressLength != IPv4SocketAddressSize)
+                return null;
+
+            long address = (long)(
+                               (socketAddress[4] & 0x000000FF) |
+                               (socketAddress[5] << 8 & 0x0000FF00) |
+                               (socketAddress[6] << 16 & 0x00FF0000) |
+                               (socketAddress[7] << 24)
+                           ) & 0x00000000FFFFFFFF;
+
+            return new IPAddress(address);
+        }
+
+        private static unsafe IPAddress CreateIPv6Address(byte* socketAddress, int addressLength)
+        {
+            const int IPv6SocketAddressSize = 28;
+            const int IPv6AddressBytes = 16;
+
+            if (addressLength != IPv6SocketAddressSize)
+                return null;
+
+            byte[] address = new byte[IPv6AddressBytes];
+            for (int i = 0; i < address.Length; i++)
+            {
+                address[i] = socketAddress[i + 8];
+            }
+
+            long scope = (long)((socketAddress[27] << 24) +
+                                (socketAddress[26] << 16) +
+                                (socketAddress[25] << 8) +
+                                (socketAddress[24]));
```

Comment:
Once CreateIPv6Address is changed to take a span, looks like this could be changed to something like:
```C#
long scope = BinaryPrimitives.ReadUInt32LittleEndian(socketAddress.Slice(24));
```
at which point the whole method (after factoring in my other comments) could basically be:
```C#
private static IPAddress CreateIPv6Address(Span<byte> socketAddress) =>
    new IPAddress(socketAddress.Slice(8, IPAddressParserStatics.IPv6AddressBytes)),
    BinaryPrimitives.ReadUInt32LittleEndian(socketAddress.Slice(24));
```
or something along those lines.

Reply:
I just saw that all those optimisations are already done in SocketAddressPal, I'll reuse them instead